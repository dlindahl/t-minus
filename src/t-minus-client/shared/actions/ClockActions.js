import ClockActions from '../actionTypes/ClockActionTypes';
import emptyObj from 'empty/object';
import formatDuration from '../util/formatDuration';
import partial from 'lodash/partial';
import present from 'present';

export const IntervalValue = 50;
let interval = null;

function calculateDuration(mode, elapsedTime, maxDuration) {
  switch(mode) {
    case ClockActions.CLOCK_MODE_TIMER:
      const timeRemaining = maxDuration - elapsedTime;
      const percentComplete = Math.min(elapsedTime / maxDuration, 1);
      return [formatDuration(timeRemaining), timeRemaining, percentComplete];
    default:
      return [formatDuration(elapsedTime), null, 1]
  }
}

function determineTimerElapsed(mode, elapsedTime, maxDuration) {
  switch(mode) {
    case ClockActions.CLOCK_MODE_TIMER:
      return elapsedTime > maxDuration;
    default:
      return false;
  }
}

export function clockTick(dispatch, getState) {
  const { clock } = getState();
  if(!clock.running) {
    return;
  }
  const currentTick = present();
  const delta = currentTick - clock.lastTick;
  const elapsedTime = clock.elapsedTime + delta;
  const [duration, timeRemaining, percentComplete] = calculateDuration(clock.mode, elapsedTime, clock.timerValue);
  const hasTimerElapsed = determineTimerElapsed(clock.mode, elapsedTime, clock.timerValue);
  dispatch({
    type: ClockActions.CLOCK_TICK,
    payload: {
      currentTick,
      delta,
      duration,
      elapsedTime,
      hasTimerElapsed,
      mode: clock.mode,
      percentComplete,
      timeRemaining
    }
  });
}

export function pauseClock() {
  return {
    type: ClockActions.CLOCK_PAUSED,
    payload: emptyObj
  };
}

export function resetClock() {
  return (dispatch, getState) => {
    const { clock } = getState();
    clearInterval(interval);
    interval = null;
    const [duration, timeRemaining, percentComplete] = calculateDuration(clock.mode, 0, clock.timerValue);
    return dispatch({
      type: ClockActions.CLOCK_RESET,
      payload: {
        duration,
        timeRemaining,
        timerValue: clock.timerValue,
        percentComplete
      }
    });
  }
}

export function startClock() {
  return (dispatch, getState) => {
    const { clock } = getState();
    const [duration, timeRemaining, percentComplete] = calculateDuration(clock.mode, clock.elapsedTime, clock.timerDuration);
    dispatch({
      type: ClockActions.CLOCK_STARTED,
      payload: {
        duration,
        percentComplete,
        startedAt: present(),
        timeRemaining
      }
    });
    interval = setInterval(partial(clockTick, dispatch, getState), IntervalValue);
  };
}

export function changeTimerValue(timerValue) {
  return (dispatch, getState) => {
    const { clock } = getState();
    const [duration, timeRemaining, percentComplete] = calculateDuration(clock.mode, 0, timerValue);
    return dispatch({
      type: ClockActions.CLOCK_TIMER_CHANGED,
      payload: { duration, percentComplete, timeRemaining, timerValue }
    });
  }
}

export function stopwatchMode() {
  const mode = ClockActions.CLOCK_MODE_STOPWATCH;
  const [duration, timeRemaining, percentComplete] = calculateDuration(mode, 0);
  return {
    type: ClockActions.CLOCK_MODE_CHANGED,
    payload: {
      duration,
      mode,
      percentComplete,
      timeRemaining
    }
  }
}

export function timerMode() {
  return (dispatch, getState) => {
    const { clock } = getState();
    const mode = ClockActions.CLOCK_MODE_TIMER;
    const [duration, timeRemaining, percentComplete] = calculateDuration(mode, 0, clock.timerValue);
    return dispatch({
      type: ClockActions.CLOCK_MODE_CHANGED,
      payload: {
        duration,
        mode,
        percentComplete,
        timeRemaining
      }
    });
  }
}

export function toggleClock() {
  return (dispatch, getState) => {
    const clock = getState().clock
    if(clock.startedAt == null) {
      startClock()(dispatch, getState);
    } else if(clock.running) {
      dispatch(pauseClock());
    } else {
      dispatch(unpauseClock());
    }
  };
}

export function unpauseClock() {
  return {
    type: ClockActions.CLOCK_UNPAUSED,
    payload: {}
  };
}
