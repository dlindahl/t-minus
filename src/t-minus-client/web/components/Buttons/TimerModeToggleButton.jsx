import AccessAlarm from 'react-icons/lib/md/access-alarm';
import AppNavButton from './AppNavButton';
import { bindActionCreators } from 'redux';
import ClockActions from '../../../shared/actionTypes/ClockActionTypes';
import { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { timerMode } from '../../../shared/actions/ClockActions';

function getState(state) {
  return {
    active: state.clock.mode === ClockActions.CLOCK_MODE_TIMER
  };
}

function getActions(dispatch) {
  return bindActionCreators({ timerMode }, dispatch);
}

@connect(getState, getActions)
export default class TimerModeToggleButton extends Component {
  constructor(props) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
  }
  handleClick() {
    this.props.timerMode();
  }
  render() {
    return (
      <AppNavButton active={this.props.active} onClick={this.handleClick}>
        <AccessAlarm/>
      </AppNavButton>
    );
  }
}

TimerModeToggleButton.propTypes = {
  active: PropTypes.bool
};

TimerModeToggleButton.defaultProps = {
  active: false
};
