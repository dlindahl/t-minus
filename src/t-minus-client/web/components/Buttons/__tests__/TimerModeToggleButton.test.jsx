import ClockActions from '../../../../shared/actionTypes/ClockActionTypes'
import configureStore from 'redux-mock-store'
import emptyObj from 'empty/object'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import TimerModeToggleButton from '../TimerModeToggleButton'

const STATES = {
  default: {
    clock: emptyObj,
    display: emptyObj
  },
  timerMode: {
    clock: {
      mode: ClockActions.CLOCK_MODE_TIMER
    },
    display: emptyObj
  }
}
const store = configureStore()

describe('TimerModeToggleButton', () => {
  it('renders the component', () => {
    const component = renderer.create(
      <Provider store={store(STATES.default)}>
        <TimerModeToggleButton/>
      </Provider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders an active component when the clock is in timer mode', () => {
    const component = renderer.create(
      <Provider store={store(STATES.timerMode)}>
        <TimerModeToggleButton/>
      </Provider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('enables stopwatch mode when clicked', () => {
    const timerMode = jest.fn()
    const component = renderer.create(
      <Provider store={store(STATES.default)}>
        <TimerModeToggleButton.WrappedComponent
          timerMode={timerMode}
        />
      </Provider>
    )
    const tree = component.toJSON()
    tree.props.onClick()
    expect(timerMode).toHaveBeenCalled()
  })
})
