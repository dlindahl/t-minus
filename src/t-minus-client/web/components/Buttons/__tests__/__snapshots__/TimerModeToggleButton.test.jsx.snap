exports[`TimerModeToggleButton renders an active component when the clock is in timer mode 1`] = `
<button
  disabled={false}
  onClick={[Function]}
  style={
    Object {
      "background": "#34495e",
      "border": 0,
      "color": "#ecf0f1",
      "cursor": "pointer",
      "fontSize": "2rem",
      "height": 50,
      "lineHeight": 1,
      "padding": 0,
      "width": 50,
    }
  }>
  <svg
    fill="currentColor"
    fit={true}
    height="1em"
    preserveAspectRatio="xMidYMid meet"
    style={
      Object {
        "verticalAlign": "top",
      }
    }
    viewBox="0 0 40 40"
    width="1em">
    <g>
      <path
        d="m20 33.4q4.8 0 8.2-3.5t3.4-8.3-3.4-8.2-8.2-3.4-8.2 3.4-3.4 8.2 3.4 8.3 8.2 3.5z m0-26.8q6.3 0 10.6 4.5t4.4 10.5-4.4 10.6-10.6 4.4-10.6-4.4-4.4-10.6 4.4-10.5 10.6-4.5z m0.9 6.8v8.7l6.6 3.9-1.2 2.1-7.9-4.7v-10h2.5z m-7.8-7.8l-7.6 6.4-2.1-2.5 7.6-6.4z m23.5 3.9l-2.1 2.6-7.6-6.5 2.1-2.5z" />
    </g>
  </svg>
</button>
`;

exports[`TimerModeToggleButton renders the component 1`] = `
<button
  disabled={false}
  onClick={[Function]}
  style={
    Object {
      "background": "transparent",
      "border": 0,
      "color": "#bdc3c7",
      "cursor": "pointer",
      "fontSize": "2rem",
      "height": 50,
      "lineHeight": 1,
      "padding": 0,
      "width": 50,
    }
  }>
  <svg
    fill="currentColor"
    fit={true}
    height="1em"
    preserveAspectRatio="xMidYMid meet"
    style={
      Object {
        "verticalAlign": "top",
      }
    }
    viewBox="0 0 40 40"
    width="1em">
    <g>
      <path
        d="m20 33.4q4.8 0 8.2-3.5t3.4-8.3-3.4-8.2-8.2-3.4-8.2 3.4-3.4 8.2 3.4 8.3 8.2 3.5z m0-26.8q6.3 0 10.6 4.5t4.4 10.5-4.4 10.6-10.6 4.4-10.6-4.4-4.4-10.6 4.4-10.5 10.6-4.5z m0.9 6.8v8.7l6.6 3.9-1.2 2.1-7.9-4.7v-10h2.5z m-7.8-7.8l-7.6 6.4-2.1-2.5 7.6-6.4z m23.5 3.9l-2.1 2.6-7.6-6.5 2.1-2.5z" />
    </g>
  </svg>
</button>
`;
